multibody/kinematic/kinematic_evaluator.cc

```diff
@@ -0,0 +1,125 @@
+#include "multibody/kinematic/kinematic_evaluator.h"
+
+using drake::multibody::MultibodyPlant;
+using drake::systems::Context;
+using drake::MatrixX;
+using drake::VectorX;
+
+namespace dairlib {
+namespace multibody {
+
+template <typename T>
+KinematicEvaluator<T>::KinematicEvaluator(const MultibodyPlant<T>& plant,
+    int length)
+    : plant_(plant),
+      length_(length) {
+  std::vector<int> all_inds;
+  all_active_ = true;
+  for (int i = 0; i < length; i++) {
+    all_inds.push_back(i);
+  }
+  set_active_inds(all_inds);
+}
+
+template <typename T>
+VectorX<T> KinematicEvaluator<T>::EvalActive(
+    const Context<T>& context) const {
+  // TODO: With Eigen 3.4, can slice by (active_inds_);
+
+  auto phi_full = EvalFull(context);
+  if (all_active_) {
+    return phi_full;
+  }
+
+  VectorX<T> phi(num_active_);
+  for (int i = 0; i < num_active_; i++) {
+    phi(i) = phi_full(active_inds_.at(i));
+  }
+  return phi;
+}
+
+template <typename T>
+VectorX<T> KinematicEvaluator<T>::EvalActiveTimeDerivative(
+    const Context<T>& context) const {
+  // TODO: With Eigen 3.4, can slice by (active_inds_);
+  auto phidot_full = EvalFullTimeDerivative(context);
+  if (all_active_) {
+    return phidot_full;
+  }
+
+  VectorX<T> phidot(num_active_);
+  for (int i = 0; i < num_active_; i++) {
+    phidot(i) = phidot_full(active_inds_.at(i));
+  }
+  return phidot;
+}
+
+template <typename T>
+MatrixX<T> KinematicEvaluator<T>::EvalActiveJacobian(
+      const Context<T>& context) const {
+  // TODO: With Eigen 3.4, can slice by (active_inds_, all);
+  auto J_full = EvalFullJacobian(context);
+  if (all_active_) {
+    return J_full;
+  }
+
+  MatrixX<T> J_active(num_active_, J_full.cols());
+  for (int i = 0; i < num_active_; i++) {
+    J_active.row(i) = J_full.row(active_inds_.at(i));
+  }
+
+  // Extract active rows only
+  return J_active;
+}
+
+template <typename T>
+VectorX<T> KinematicEvaluator<T>::EvalActiveJacobianDotTimesV(
+      const Context<T>& context) const {
+  // TODO: With Eigen 3.4, can slice by (active_inds_);
+  auto Jdot_v_full = EvalFullJacobianDotTimesV(context);
+  if (all_active_) {
+    return Jdot_v_full;
+  }
+
+  VectorX<T> Jdot_v(num_active_);
+  for (int i = 0; i < num_active_; i++) {
+    Jdot_v(i) = Jdot_v_full(active_inds_.at(i));
+  }
+  return Jdot_v;
+}
+
+template <typename T>
+VectorX<T> KinematicEvaluator<T>::EvalFullTimeDerivative(
+    const Context<T>& context) const {
+  return EvalFullJacobian(context) * plant_.GetVelocities(context);
+}
+
+template <typename T>
+void KinematicEvaluator<T>::set_active_inds(std::vector<int> active_inds) {
+  // Check active_inds [0, length()] bounds
+  for (auto& i : active_inds) {
+    DRAKE_DEMAND(i >= 0);
+    DRAKE_DEMAND(i < num_full());
+  }
+  // TODO: should probably check for uniqueness
```

Comment:
I agree that we should check uniqueness. By a quick google search, this piece of code remove the duplicates: 
```
std::sort( vec.begin(), vec.end() );
vec.erase( std::unique( vec.begin(), vec.end() ), vec.end() );
```

Reply:
I actually had something in here that would do this, then removed it. I'll re-add. I prefer throwing an error on non-unique solutions than secretly changing the user's request.

Reply:
Yes throwing an error might be better